api/app/src/main/kotlin/packit/service/RolePermissionService.kt

```diff
@@ -81,4 +101,90 @@ class BaseRolePermissionService(
 
         return rolePermissionsToAdd
     }
+
+    override fun updatePacketReadPermissionOnRoles(
+        rolesToAdd: List<Role>,
+        rolesToRemove: List<Role>,
+        packetGroupName: String,
+        packetId: String?,
+    )
+    {
+        val permission = permissionService.getByName("packet.read")
+        val packet = packetId?.let { packetService.getPacket(it) }
+        val packetGroup = if (packetId == null)
+        {
+            packetGroupService.getPacketGroupByName(packetGroupName)
+        } else
+        {
+            null
+        }
+
+        removePermissionFromRoles(
+            rolesToRemove,
+            permission,
+            packet,
+            packetGroup
+        )
+        addPermissionToRoles(
+            rolesToAdd,
+            permission,
+            packet,
+            packetGroup
+        )
+        roleRepository.saveAll(rolesToAdd)
+    }
+
+    override fun addPermissionToRoles(
+        roles: List<Role>,
+        permission: Permission,
+        packet: Packet?,
+        packetGroup: PacketGroup?,
+        tag: Tag?,
+    )
+    {
+        for (role in roles)
+        {
+            val rolePermission = RolePermission(
+                role = role,
+                permission = permission,
+                packet = packet,
+                packetGroup = packetGroup,
+                tag = tag,
+            )
+            if (role.rolePermissions.any { it == rolePermission })
+            {
+                throw PackitException("rolePermissionAlreadyExists", HttpStatus.BAD_REQUEST)
```

Comment:
Alternatively, we could make this method idempotent and do nothing if the entry already exists and not throw an error. I'd be fine with either approach.

Reply:
yeah that may be an option. Im personally more of a fan of failing just because the user will know if they have messed something up, and not get confused like thinking they have added a role or something.